### PR TITLE
Allow to read public contents without login through API not working #54

### PR DESCRIPTION
**Pull request by gaurav-handysolver**
We added two functions handlePagination() and returnPagination() in AuthBaseController file.

Note: We noticed sometimes data not getting because of readable() function attached with the query in calendarController. so, we removed that function in calendarController

#54

### CODE REVIEW

**Comment by donni106** [comment]
All right, works again, thanks.
But can you please return only public data when there is no bearer token? I see the same data (visibility=0 and visibility=1) no matter if i pass a bearer token or not. I guess visibility is the correct key to identify the public data, right?

**Comment by donni106** [comment]
It is implemented for spaces already, right? `$query = Space::find()->where(['visibility' => Space::VISIBILITY_ALL]);`
But not for posts and calendar entries? @gaurav-handysolver

**Comment by gaurav-handysolver** [comment]
Hi @donni106 
Yes, we can modify the query so it will return only public data if we will not pass the bearer token.

So, should we change the query?

**Comment by donni106** [comment]
Yes please, 
- without bearer token -> only public data
- with bearer token -> all data

**Comment by gaurav-handysolver** [comment]
Hi @donni106 
Should this logic (private and public) will apply on '/calendar/container/container-id' and 'calendar/entry/entry-id' endpoints?

**Comment by donni106** [comment]
> Should this logic (private and public) will apply on '/calendar/container/container-id' and 'calendar/entry/entry-id' endpoints?

yes please and `/calendar`, right?
please check the three v2 endpoints for `posts` as well, if this logic needs to be applied.
please check it also for `/space` and `/space/{id}`. currently i cannot see a difference between sending a bearer token or not.

**Comment by gaurav-handysolver** [comment]
Okay @donni106 
Thanks!

File: controllers/calendar/CalendarController.php [hunk 1]
```diff
@@ -78,4 +106,17 @@ public function actionFindByContainer($containerId)
         return $this->returnPagination($query, $pagination, $results);
     }
 
+    private function checkUserIsRegistered(){
+        $authHeader = Yii::$app->request->getHeaders()->get('Authorization');
+        if (!empty($authHeader) && preg_match('/^Bearer\s+(.*?)$/', $authHeader, $matches)) {
+            $token = $matches[1];
+
+            $validData = JWT::decode($token, "NFqRh33ofXCLYQ9SvFiX3lnBa7qLl2NcMBj_gYMaTCwdcxSIqY3rYxJ2UWXiE1R0Ow0oYg4fJk9HaVGykWzFry", ['HS512']);
```
Comment (donni106) [inline controllers/calendar/CalendarController.php:114]: what does this token mean here? where does it come from? is it safe to hardcode it?

Comment (donni106) [reply controllers/calendar/CalendarController.php:114]: this PR is an urgent topic for us, @gaurav-handysolver can you please have a look here to say something about the hardcoded token, thanks

Comment (donni106) [reply controllers/calendar/CalendarController.php:114]: yes please replace the hardcoded string, thanks.

Comment (gaurav-handysolver) [reply controllers/calendar/CalendarController.php]: Hi @donni106 
Done!

File: controllers/space/SpaceController.php [hunk 2]
```diff
@@ -2,44 +2,67 @@
...
+use humhub\modules\user\models\User;
 
 class SpaceController extends AuthBaseController
 {
     public function actionFind(){
         $results = [];
-        $query = Space::find()->where(['visibility' => Space::VISIBILITY_ALL]);
+
+        //Check the requested user is guest or not
+        $user = $this->checkUserIsRegistered();
```
Comment (donni106) [inline controllers/space/SpaceController.php:18]: something seems to be still bad with the space controller:

```
{
	"name": "Exception",
	"message": "Class 'humhub\\modules\\smartVillage\\controllers\\space\\Yii' not found",
	"code": 0,
	"type": "Error",
	"file": "/var/www/localhost/htdocs/protected/humhub/modules/smart-village-app-humhub-ext/controllers/space/SpaceController.php",
	"line": 55,
	"stack-trace": [
		"#0 /var/www/localhost/htdocs/protected/humhub/modules/smart-village-app-humhub-ext/controllers/space/SpaceController.php(18): humhub\\modules\\smartVillage\\controllers\\space\\SpaceController->checkUserIsRegistered()",
		"#1 [internal function]: humhub\\modules\\smartVillage\\controllers\\space\\SpaceController->actionFind()",
		"#2 /var/www/localhost/htdocs/protected/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)",
		"#3 /var/www/localhost/htdocs/protected/vendor/yiisoft/yii2/base/Controller.php(181): yii\\base\\InlineAction->runWithParams(Array)",
		"#4 /var/www/localhost/htdocs/protected/vendor/yiisoft/yii2/base/Module.php(534): yii\\base\\Controller->runAction('find', Array)",
		"#5 /var/www/localhost/htdocs/protected/vendor/yiisoft/yii2/web/Application.php(104): yii\\base\\Module->runAction('smartVillage/sp...', Array)",
		"#6 /var/www/localhost/htdocs/protected/vendor/yiisoft/yii2/base/Application.php(392): yii\\web\\Application->handleRequest(Object(humhub\\components\\Request))",
		"#7 /var/www/localhost/htdocs/index.php(25): yii\\base\\Application->run()",
		"#8 {main}"
	]
}
```

calendar and post are working.
any idea @gaurav-handysolver?

File: controllers/space/SpaceController.php [hunk 3]
```diff
@@ -2,44 +2,67 @@
 
 namespace humhub\modules\smartVillage\controllers\space;
 
+use Firebase\JWT\JWT;
+use humhub\modules\rest\models\ConfigureForm;
 use humhub\modules\smartVillage\components\AuthBaseController;
 use humhub\modules\space\models\Space;
 use humhub\modules\rest\definitions\SpaceDefinitions;
+use humhub\modules\user\models\User;
```
Comment (donni106) [inline controllers/space/SpaceController.php:10]:
```
use Yii;
``` 

missing here?

File: controllers/space/SpaceController.php [hunk 4]
```diff
@@ -2,44 +2,67 @@
...
-        $query = Space::find()->where(['visibility' => Space::VISIBILITY_ALL]);
+
+        //Check the requested user is guest or not
+        $user = $this->checkUserIsRegistered();
+        if($user == false){
+            $query = Space::find()->where(['visibility' => Space::VISIBILITY_ALL]);
+        }else{
+            $query = Space::find()->where(['visibility' => Space::VISIBILITY_REGISTERED_ONLY]);
+
+        }
```
Comment (donni106) [inline controllers/space/SpaceController.php:24]: now if there is no user, all spaces with `"visibility": 2` are returned and if there is a user only the spaces with `"visibility": 1,`. but if the user is logged in there should be all spaces returned, 1s and 2s, you know? can you please adjust this as well @gaurav-handysolver.

File: controllers/space/SpaceController.php [hunk 5]
```diff
@@ -2,44 +2,67 @@
...
-        if(!isset($checkSpace)){
+        if(!isset($space)){
             return $this->returnError(400,"Space not found!");
         }
 
-        if(!isset($checkVisibility)){
-            return $this->returnError(400,"Guest user cannot read this space data");
+        $user = $this->checkUserIsRegistered();
+        if($user == false){
+            if($space->visibility == Space::VISIBILITY_REGISTERED_ONLY){
```
Comment (donni106) [inline controllers/space/SpaceController.php:44]: this should also apply for `VISIBILITY_NONE` or am i wrong? guest users should not be able to see private spaces

File: controllers/space/SpaceController.php (hunk 3 above)
Comment (gaurav-handysolver) [reply controllers/space/SpaceController.php:10]: Sorry, my bad

File: controllers/space/SpaceController.php (hunk 4 above)
Comment (donni106) [reply controllers/space/SpaceController.php:24]: when i am thinking about different cases i guess we need every space except the ones with `"visibility": 0` in the second case with a user, right?

Comment (donni106) [reply controllers/space/SpaceController.php:24]: the three possible states: https://github.com/humhub/humhub/blob/10c8ee6702b1ab1196db6cb7366b3ed18e003493/protected/humhub/modules/space/models/Space.php#L76-L78

so we need "where not NONE" right?

File: controllers/space/SpaceController.php (hunk 5 above)
Comment (gaurav-handysolver) [reply controllers/space/SpaceController.php]: Yes, you are right and add one more condition in this logic

File: controllers/space/SpaceController.php (hunk 4 above)
Comment (gaurav-handysolver) [reply controllers/space/SpaceController.php]: Done

File: controllers/space/SpaceController.php (hunk 2 above)
Comment (donni106) [reply controllers/space/SpaceController.php:19]: `use Yii` was missing

**Review by donni106 — APPROVED** [review]

**Pull request merged**